### PR TITLE
KAFKA-14367; Add `DescribeGroups` to the new `GroupCoordinator` interface

### PR DESCRIPTION
**Pull request by dajac**
This patch adds `describeGroup` to the new `GroupCoordinator` interface and updates `KafkaApis` to use it.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: I see we changed this to be plural -- was the reasoning because of the API?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: nit: but does this need to be 3 lines?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: did we want a trace log here too?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Not a huge deal but was wondering how you decided what to include in the newGroupCoordinator method vs what to keep outside of it. The below lines about exception being null could be included right?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: That's right.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: let me put it on one line.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: added it.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: No, it could not because `newGroupCoordinator.describeGroup` could actually return a future containing an exception (expected or unexpected) so we have to handle this case here as well.

Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Ah I see. Thanks for clarifying

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Did we want to refactor to include the new SendResponse method from the SyncGroup PR?

Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Yes, will do when that PR is merged.

**Review by jolshan — APPROVED** [review]
lgtm -- once the other pr is merged and we add the helper.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: I wonder if it would be simpler to pass through all groups that we want to describe. Then we just have a single future.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: A little outside the scope of this patch, but I think we can improve the name here. The class `DescribeGroupsResponse` has both `forError` and `fromError` and it's not really clear what the differences are from the naming. Since this is a group-level error, perhaps we can emphasize that in the name. Maybe we could call it `groupError` or something like that?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Any idea why the original semantics were to handle each group one by one?

Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: There is not particular reason, I guess. @hachikuji The reason I kept it like this is that we will need to fan out any way with the new controller as we will have multiple event loops. I suppose that we could say that this is an internal implementation details so not exposing it in the interface would make sense. What do you think?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Sure. Let me clean this. I wonder if we should also remove `fromError` and push that code to `DescribeGroupsRequest.getErrorResponse` in order to be consistent with the other requests.

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Yeah, I was wondering why we didn't use `getErrorResponse`.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I don't have a strong opinion, but I slightly prefer keeping `KafkaApis` simple and having the complexity in the coordinator implementation. I also wonder if that makes optimization easier. For example, even if we have fanout to the respective coordinators, it may still be the case that multiple groups are handled by the same coordinator. Perhaps that would mean fewer messages on the respective queues? Not sure.

Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Yeah, I think that I am with you on this point. Keeping `KafkaApis` simple makes sense. Let me give it a shot tomorrow.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I did the refactor separately: https://github.com/apache/kafka/pull/12970.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I have refactored the code to address this.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: https://github.com/apache/kafka/pull/12970 was merged.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala:83]: nit: line here -- not sure what the expected format is though 😅

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala
Comment (jolshan) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala:194]: do we want to rename this to describeGroups since we have multiple group IDs?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:2076]: was this change intentional?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:2116]: I think this is a good addition to test. 👍

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (dajac) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:2076]: yes.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala:83]: reverted this change.

**Review by jolshan — APPROVED** [review]
lgtm thanks!

**Review by hachikuji — APPROVED** [review]
Thanks, LGTM

**Pull request merged**
